pack-redundant: optimize for memory and cpu time

sunchao9 opened the pull request:
pack-redundant: optimize for memory and cpu time

When call pack-redundant with "--all", if there are too much local packs and
too much redundant objects within them, the **minimize** step may eat up the
memory and cpu because it would produces a huge-big-size "pll" list in the
heap.

We could reproduct the case that `pack-redundant --all` will eat up cpu-time
by this script:

> thanks to Jiang Xin to help finishing this test script

```bash

repo="$(pwd)/test.git"
work="$(pwd)/test"
i=1
max=1000

git init -q --bare "$repo"
git --git-dir="$repo" config gc.auto 0
git --git-dir="$repo" config transfer.unpackLimit 100
git clone -q "$repo" "$work" 2>/dev/null

while :; do
    cd "$work"
    echo "loop $i: date +%s" >$i
    git add -A
    git commit -q -sm "loop $i"
    git push -q origin HEAD:master
    printf "\rCreate pack %4d/%d\t" $i $max

    cd "$repo"
    git repack -q

    if test $(($i % 2)) -eq 0; then
        git repack -aq
        pack=$(ls -t $repo/objects/pack/*.pack | head -1)
        touch "${pack%.pack}.keep"
    fi

    i=$((i+1))
    if test $i -gt $max; then break; fi
done
printf "\ndone\n"
```

Change the strategy of **minimize** could solve this:

- 1. Get the unique and non_uniqe packs, add the unique packs to the **min**
list.
- 2. Remove the objects of the unique packs from the non_unique packs, Then the
left objects in the non_unique packs will have the same feature: every object
will exists at least two different packs.
- 3. Sort the non_unique packs by the objects' size, add the first pack to the
**min** list, and remove it from the non_unique packs.
- 4. Delete the objects of the first pack from the other packs, and delete the
packs with zero objects.
- 5. Repate 3 And 4 until no non_unique packs left.

With the strategy for **minimize** above, the **min** list is the collection
with smallest size. And **minimize** need not to get the permutations for the
missing objects from the unique set, which need less memory and cpu time.

Signed-off-by: sunchao <sunchao9@huawei.com>

jiangxin commented:
改了下脚本，以便在 mac 上能执行

    #!/bin/sh

    repo="$(pwd)/test.git"
    work="$(pwd)/test"
    i=1
    max=1000

    git init -q --bare "$repo"
    git --git-dir="$repo" config gc.auto 0
    git --git-dir="$repo" config transfer.unpackLimit 100
    git clone -q "$repo" "$work" 2>/dev/null

    while :; do
        cd "$work"
        echo "loop $i: date +%s" >$i
        git add -A
        git commit -q -sm "loop $i"
        git push -q origin HEAD:master
        printf "\rCreate pack %4d/%d\t" $i $max

        cd "$repo"
        git repack -q

        if test $(($i % 2)) -eq 0; then
            git repack -aq
            pack=$(ls -t $repo/objects/pack/*.pack | head -1)
            touch "${pack%.pack}.keep"
        fi

        i=$((i+1))
        if test $i -gt $max; then break; fi
    done
    printf "\ndone\n"

sunchao9 commented:
> 改了下脚本，以便在 mac 上能执行
> 
> ```
> #!/bin/sh
> 
> repo="$(pwd)/test.git"
> work="$(pwd)/test"
> i=1
> max=1000
> 
> git init -q --bare "$repo"
> git --git-dir="$repo" config gc.auto 0
> git --git-dir="$repo" config transfer.unpackLimit 100
> git clone -q "$repo" "$work" 2>/dev/null
> 
> while :; do
>     cd "$work"
>     echo "loop $i: date +%s" >$i
>     git add -A
>     git commit -q -sm "loop $i"
>     git push -q origin HEAD:master
>     printf "\rCreate pack %4d/%d\t" $i $max
> 
>     cd "$repo"
>     git repack -q
> 
>     if test $(($i % 2)) -eq 0; then
>         git repack -aq
>         pack=$(ls -t $repo/objects/pack/*.pack | head -1)
>         touch "${pack%.pack}.keep"
>     fi
> 
>     i=$((i+1))
>     if test $i -gt $max; then break; fi
> done
> printf "\ndone\n"
> ```

使用这个脚本，我用 `git version 2.18.0` 在 `Ubuntu 14.04.5 LTS` 上可以复现：

#### minimize 开始，经过 cmp_two_packs 产生了missing objects，对pack队列进行一一对比，查找是否有单个pack能够包含所有的missing objects
![image](https://user-images.githubusercontent.com/14326983/48877425-14d9c800-ee3e-11e8-958b-cb4b71304bb9.png)

#### minimize 无法找到满足包含所有 missing objects 的单个pack包，进一步开始进行2-packs组合进行对比，如果2-packs组合不满足，开始3-pakcs组合.. 4-packs 组合，此时可能会耗尽内存
![image](https://user-images.githubusercontent.com/14326983/48877382-ee1b9180-ee3d-11e8-9a44-4e1165ff7a57.png)


回头我尝试下mac。：D

sunchao9 commented:
@jiangxin 目前看在 git/git 上面的CI已经pass了：  https://github.com/git/git/pull/558

jiangxin commented:
如图9个集合，使用当前算法，按照从1到9顺序，是找不出来冗余的 pack 的。
> ![chart](https://user-images.githubusercontent.com/183860/49627519-42b43480-fa1a-11e8-9d0b-e84e6e17111a.jpg)

如果把对象和packfile列一张表，如下图所示：
> ![tables](https://user-images.githubusercontent.com/183860/49627520-4942ac00-fa1a-11e8-87f3-2bef9ff30faa.jpg)
先筛掉包含唯一对象的第九列，和对应的行（objects），剩下的对象都至少有两个 packfile 对应。然后根据 packfile 包含表中剩余 objects 的数量的多寡排序，从大到小逐一筛掉，剩下的空 pack 文件就是冗余的 packfile。

jiangxin commented:
> > 改了下脚本，以便在 mac 上能执行
> 回头我尝试下mac。：D

这个复现问题的脚本，直接放在 commit log 中就更好了，按照 markdown 语法缩进。

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -421,14 +421,52 @@ static inline off_t pack_set_bytecount(struct pack_list *pl)
 }
 
-static void minimize(struct pack_list **min)
+static int sort_packs_size(const void *a_, const void *b_)
```
这个方法是比较两个list，方法重命名为 `cmd_pack_list` 为好。

sunchao9 replied:
OK，我修改为 `cmp_pack_list`

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -421,14 +421,52 @@ static inline off_t pack_set_bytecount(struct pack_list *pl)
+	for (n = 0, p = *pl; p; p = p->next)
+		ary[n++] = p;
+
+	QSORT(ary, n, sort_packs_size);
```
> 方法 `range_packs_by_objects_size` 会更改入参 pack_list **pl 本身吧，那么是否还需要一个 ary 参数？是不是直接对 pk 做 QSORT？

QSORT 是针对数组的，不能对链表排序，实现是对的。

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -421,14 +421,52 @@ static inline off_t pack_set_bytecount(struct pack_list *pl)
+	return 1;
+}
+
+static void range_packs_by_objects_size(struct pack_list **pl)
```
方法名是不是改成 `sort_pack_list_by_size`， 因为 range 好像是一个迭代器，类似于 golang 里面的 range。

sunchao9 replied:
OK

sunchao9 commented:
@jiangxin 
git的CI测试通过： https://github.com/git/git/pull/564

jiangxin reviewed (commented):
rebase to v2.20.0 and see my new comments

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -453,41 +491,46 @@ static void minimize(struct pack_list **min)
+	unique_objects = llist_copy(all_objects);
+	llist_sorted_difference_inplace(unique_objects, missing);
+
+	/* remove all the ignored objects and unique pack objects from the non_unique packs */
```
为什么不把 uniq pack包 添加到 min 中呢？例如：

    min = unique;

sunchao9 replied:
应该需要把 min 先指向 unique

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -453,41 +491,46 @@ static void minimize(struct pack_list **min)
+
+	while (non_unique) {
+		/* sort the non_unique packs by all_objects's size */
+		sort_pack_list_by_size(&non_unique);
```
排序结果是升序的吧? 那么是不是第一个对象的 all_objects->size ==0 ?

sunchao9 replied:
排序有问题，我排成倒序了

sunchao9 replied:
`cmp_pack_list` 需要修改为:

```c
static int cmp_pack_list(const void *a_, const void *b_)
{
	struct pack_list *a = *((struct pack_list **)a_);
	struct pack_list *b = *((struct pack_list **)b_);

	size_t a_size = a->all_objects->size;
	size_t b_size = b->all_objects->size;
	// let the pack with smaller all_objects sizesorted to the right
	if (a_size == b_size)
		return 0;
	else if (a_size < b_size)
		return 1;

	return -1;
}
```

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -453,41 +491,46 @@ static void minimize(struct pack_list **min)
-				memcpy(new_perm, perm, sizeof(struct pll));
-				new_perm->next = perm_ok;
-				perm_ok = new_perm;
+	unique_objects = llist_copy(all_objects);
```
将 unique_object 重命名为 `unique_pack_objects` 为好。

sunchao9 replied:
同意

jiangxin commented:
修改了测试脚本，原来的脚本生成不了 unique pack。这个脚本保证最后一个 pack 是 unique pack。
测试发现现在版本的修改，会造成 unique pack 丢失，因为 min 没有包含 unique。

    #!/bin/sh
    
    repo="$(pwd)/test.git"
    work="$(pwd)/test"
    i=1
    max=199
    
    git init -q --bare "$repo"
    git --git-dir="$repo" config gc.auto 0
    git --git-dir="$repo" config transfer.unpackLimit 0
    git clone -q "$repo" "$work" 2>/dev/null
    
    while :; do
        cd "$work"
        echo "loop $i: $(date +%s)" >$i
        git add -A
        git commit -q -sm "loop $i"
        git push -q origin HEAD:master
        printf "\rCreate pack %4d/%d\t" $i $max
    
        if test $i -ge $max; then break; fi
    
        cd "$repo"
        git repack -q
    
        if test $(($i % 2)) -eq 0; then
            git repack -aq
            pack=$(ls -t $repo/objects/pack/*.pack | head -1)
            touch "${pack%.pack}.keep"
        fi
    
        i=$((i+1))
    done
    printf "\ndone\n"

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -453,41 +491,46 @@ static void minimize(struct pack_list **min)
+		sort_pack_list_by_size(&non_unique);
+		if (non_unique->all_objects->size == 0) {
+			break;
+		}
```
如果有一个包是 unique 包的子集，则 all_objects 的size 为0. 那么在这个循环一开始就 break 了。
所以一开始就要过滤掉 `all_objects->size` 为0的 non_unique 包。

        while (non_unique) {
                /* sort the non_unique packs by all_objects's size */
                sort_pack_list_by_size(&non_unique);

                /* skip empty non_unique (subset of unique packs) */
                while (non_unique && non_unique->all_objects->size == 0) {
                        non_unique = non_unique->next;
                }   
                if (!non_unique)
                        break;

sunchao9 replied:
后面的做法是用`all_objects`最多的pack来清理其它pack中重复的object，所以之前将`all_object->size`为0的放在左边是排序错了，将最多的应该放在左边

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -453,41 +491,46 @@ static void minimize(struct pack_list **min)
+		/* remove the pack with zero size all_objects */
+		pl = non_unique->next;
+		while (pl) {
+			if (pl == NULL || pl->all_objects->size == 0) {
```
non_unique 不是 sort 过了么？除了前面会有 `all_objects->size` 为 0，后面不会有。

sunchao9 replied:
同意，错误问题同上

jiangxin reviewed (commented):
See this patch：

    diff --git a/builtin/pack-redundant.c b/builtin/pack-redundant.c
    index 785d627b93..4aa433854c 100644
    --- a/builtin/pack-redundant.c
    +++ b/builtin/pack-redundant.c
    @@ -388,7 +388,7 @@ static void sort_pack_list_by_size(struct pack_list **pl)
     static void minimize(struct pack_list **min, struct llist *ignore)
     {
     	struct pack_list *old, *pl, *unique = NULL, *non_unique = NULL;
    -	struct llist *missing, *unique_objects;
    +	struct llist *missing, *unique_pack_objects;
     
     	pl = local_packs;
     	while (pl) {
    @@ -413,38 +413,28 @@ static void minimize(struct pack_list **min, struct llist *ignore)
     		return;
     	}
     
    -	unique_objects = llist_copy(all_objects);
    -	llist_sorted_difference_inplace(unique_objects, missing);
    +	*min = unique;
    +	unique_pack_objects = llist_copy(all_objects);
    +	llist_sorted_difference_inplace(unique_pack_objects, missing);
     
     	/* remove all the ignored objects and unique pack objects from the non_unique packs */
     	pl = non_unique;
     	while (pl) {
     		llist_sorted_difference_inplace(pl->all_objects, ignore);
    -		llist_sorted_difference_inplace(pl->all_objects, unique_objects);
    +		llist_sorted_difference_inplace(pl->all_objects, unique_pack_objects);
     		pl = pl->next;
     	}
     
     	while (non_unique) {
     		/* sort the non_unique packs by all_objects's size */
     		sort_pack_list_by_size(&non_unique);
    -		if (non_unique->all_objects->size == 0) {
    -			break;
    -		}
    -
    -		/* remove the pack with zero size all_objects */
    -		pl = non_unique->next;
    -		while (pl) {
    -			if (pl == NULL || pl->all_objects->size == 0) {
    -				break;
    -			}
    -			pl = pl->next;
    -		}
     
    -		while (pl) {
    -			old = pl;
    -			pl = pl->next;
    -			free(old);
    +		/* skip empty non_unique (subset of unique packs) */
    +		while (non_unique && non_unique->all_objects->size == 0) {
    +			non_unique = non_unique->next;
     		}
    +		if (!non_unique)
    +			break;
     
     		pack_list_insert(min, non_unique);
     		pl = non_unique->next;

Commit:
pack-redundant: new algorithm to find min packs

When calling `git pack-redundant --all`, if there are too many local
packs and too many redundant objects within them, the too deep iteration
of `get_permutations` will exhaust all the resources, and the process of
`git pack-redundant` will be killed.

The following script could create a repository with too many redundant
packs, and running `git pack-redundant --all` in the `test.git` repo
will die soon.

    #!/bin/sh

    repo="$(pwd)/test.git"
    work="$(pwd)/test"
    i=1
    max=199

    if test -d "$repo" || test -d "$work"; then
    	echo >&2 "ERROR: '$repo' or '$work' already exist"
    	exit 1
    fi

    git init -q --bare "$repo"
    git --git-dir="$repo" config gc.auto 0
    git --git-dir="$repo" config transfer.unpackLimit 0
    git clone -q "$repo" "$work" 2>/dev/null

    while :; do
        cd "$work"
        echo "loop $i: $(date +%s)" >$i
        git add $i
        git commit -q -sm "loop $i"
        git push -q origin HEAD:master
        printf "\rCreate pack %4d/%d\t" $i $max
        if test $i -ge $max; then break; fi

        cd "$repo"
        git repack -q
        if test $(($i % 2)) -eq 0; then
            git repack -aq
            pack=$(ls -t $repo/objects/pack/*.pack | head -1)
            touch "${pack%.pack}.keep"
        fi
        i=$((i+1))
    done
    printf "\ndone\n"

To get the `min` unique pack list, we can replace the iteration in
`minimize` function with a new algorithm, and this could solve this
issue:

1. Get the unique and non_uniqe packs, add the unique packs to the
   `min` list.

2. Remove the objects of unique packs from non_unique packs, then each
   object left in the non_unique packs will have at least two copies.

3. Sort the non_unique packs by the objects' size, more objects first,
   and add the first non_unique pack to `min` list.

4. Drop the duplicated objects from other packs in the ordered
   non_unique pack list, and repeat step 3.

Original PR and discussions: https://github.com/jiangxin/git/pull/25

Signed-off-by: Sun Chao <sunchao9@huawei.com>
Signed-off-by: Jiang Xin <worldhello.net@gmail.com>

Commit:
pack-redundant: remove unused functions

Remove unused functions to find `min` packs, such as `get_permutations`,
`pll_free`, etc.

Signed-off-by: Sun Chao <sunchao9@huawei.com>
Signed-off-by: Jiang Xin <worldhello.net@gmail.com>

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -446,48 +484,55 @@ static void minimize(struct pack_list **min)
+		old = non_unique;
+		pl = NULL;
+		while (old) {
+			pl = old->next;
```
后面的521行 `old=old->next`和本行，让 pl 每次向后跳2次。

sunchao9 replied:
已经重新修改了代码逻辑

jiangxin commented on builtin/pack-redundant.c:
```diff
@@ -446,48 +484,55 @@ static void minimize(struct pack_list **min)
+		while (old) {
+			pl = old->next;
+			if (pl == NULL || pl->all_objects->size == 0) {
+				old->next = NULL;
```
这里设置 old->next 为 NULL，后面的释放内存的循环就没有道理了。

sunchao9 replied:
同意，后面释放的操作不需要了

jiangxin commented:
Hi @sunchao9,

Amend your commits in branch `sc/pack-redundant`.

jiangxin commented:
We may not delete function `llist_free` in commit 2.

jiangxin commented:
Patches have been sent to git mailing list, see:

https://public-inbox.org/git/20181218095829.20092-1-worldhello.net@gmail.com/T/#t

sunchao9 commented:
> Patches have been sent to git mailing list, see:
> 
> https://public-inbox.org/git/20181218095829.20092-1-worldhello.net@gmail.com/T/#t

我增加了一个patch，修改了下可能导致少统计pack的问题，请看下:

```diff
diff --git a/builtin/pack-redundant.c b/builtin/pack-redundant.c
index d0ff237..63e3291 100644
--- a/builtin/pack-redundant.c
+++ b/builtin/pack-redundant.c
@@ -422,7 +422,7 @@ static void minimize(struct pack_list **min, struct llist *ignore)
                pl = pl->next;
        }
 
-       while ((pl = non_unique)) {
+       while (non_unique) {
                /* sort the non_unique packs, greater size of all_objects first */
                sort_pack_list(&non_unique);
                if (non_unique->all_objects->size == 0)
@@ -430,6 +430,7 @@ static void minimize(struct pack_list **min, struct llist *ignore)
 
                pack_list_insert(min, non_unique);
 
+               pl = non_unique;
                while ((pl = pl->next)) {
                        if (pl->all_objects->size == 0)
                                break;
```

jiangxin commented:
Junio 已经创建特性分支并合入 `pu` 分支，关注上游 `what's cooking in git.git` 邮件，关注进展。
